man/man1/plookup.1
.\" Automatically generated by Pandoc 2.14.2
.\"
.TH "plookup" "1" "" "2021-08-17" "Open PMIx"
.hy
.SH NAME
.PP
plookup - Request the value of one or more keys
.SH SYNOPSIS
.IP
.nf
\f[C]
plookup [ options ] [ keys ]
\f[R]
.fi
.SH DESCRIPTION
.PP
\f[C]plookup\f[R] will request the value of the specified keys.
This requires that \f[C]plookup\f[R] connect to an appropriate server -
the tool will automatically attempt to do so, but options are provided
to assist its search or to direct it to a specific server.
Reserved keys must be provided in their string form - e.g.,
\[lq]pmix.job.term.status\[rq] as opposed to the key\[cq]s name of
\[lq]PMIX_JOB_TERM_STATUS\[rq].
.SH OPTIONS
.PP
\f[C]plookup\f[R] supports the following options:
.IP
.nf
\f[C]
-h|--help                This help message
-p|--pid <arg0>          Specify starter pid
-t|--timeout             Max number of seconds to wait for data to become
                         available
-v|--verbose             Be Verbose
-w|--wait                Wait for data to be available
\f[R]
.fi
.SH AUTHORS
.PP
The OpenPMIx maintainers \[en] see https://github.com/openpmix/openpmix
or the file \f[C]AUTHORS\f[R].


NEWS
Copyright (c) 2015-2020 Intel, Inc.  All rights reserved.
Copyright (c) 2017-2020 IBM Corporation.  All rights reserved.
Copyright (c) 2021      Nanook Consulting.  All rights reserved.
Copyright (c) 2021      Amazon.com, Inc. or its affiliates.
                        All Rights reserved.
$COPYRIGHT$

Additional copyrights may follow

$HEADER$

===========================================================================

This file contains the main features as well as overviews of specific
bug fixes (and other actions) for each version of PMIx since
version 1.0.

As more fully described in the "Software Version Number" section in
the README file, PMIx typically maintains two separate version
series simultaneously - the current release and one that is locked
to only bug fixes.  Since these series are semi-independent of each
other, a single NEWS-worthy item might apply to different series. For
example, a bug might be fixed in the master, and then moved to
multiple release branches.

4.1.1 -- TBD
----------------------
**** NOTE: As of v4.1.1, PMIx includes an EXCEPTIONS file that lists
**** all deviations from the PMIx Standard. This primarily includes
**** extensions that have not yet been adopted by the Standard.

**** NOTE: As of v4.1.1, PMIx no longer has a dependency on "pandoc"
**** for building and installing man pages.

 - PR #2299: Update 4.1.1 rc2
    - Update VERSION and NEWS
    - Add some spawn-specific timeout attributes
    - Resolve race condition in lost connection
    - Provide "partial_success" error when collectives not complete
    - Only conditionally decode the nspace return value when we
        are sure the spawn was successful
    - Correctly copy stone age hwloc topologies
    - Remove man page Markdown source and build dependency on pandoc
 - PR #2277: Update v4.1.1 release candidate
    - Add missing m4 file
    - Add a Standard extension value to the compliance version
    - Properly read/output stdout/err from a fork/exec'd child
    - Default to using our local_output flag
    - Cleanup compiler warnings for ancient hwloc versions
    - Prefix the output files with "pmix"
    - Ensure tools wait until all active events are processed
    - autogen.pl: ignore all excluded components
    - Don't treat inability to open shmem file as fatal
    - Avoid use of MCA params for singleton and report-uri
    - Ensure the server waits for all IOF and message events to complete
    - Restore the thread join in progress thread "stop"
    - Mark the read event as no longer active
    - Avoid blocking in the stdin read handler
    - Some cleanup of IOF output
    - Add missing .m4 files to extra_dist
    - Check for libevent minimum 2.0.21
    - Add Intel GPU component
    - Correct vendor IDs and generalize check_vendor
    - Add missing storage-related datatype support
    - Add missing storage constants
    - Improve pnet component selection
    - Cleanup the device distance computation


4.1.0 -- 29 July 2021
----------------------
**** NOTE: This release implements the complete PMIx v4.1 Standard
**** and therefore includes a number of new APIs and features. These
**** are fully documented in the official document. It also includes
**** some extensions that have not yet been included in that document.

Beyond the v4.1 modifications and additions to APIs, datatypes, attributes,
and macros, changes to the library include:

 - PR #2251: More updates from master
     - Replaced PMIx_Notify_Event with cbfunc call in errhandler to match
     - Update attribute support tables
 - PR #2248: Continue updates to support MPICH integration
     - Extend IOF outputting format to cover Hydra options.
 - PR #2246: Cleanup some IOF attributes
 - PR #2235: Cherry-pick updates from master branch
     - Default tools to outputting their IOF
     - Initalize val before get in case get isn't successful
     - Some cleanups of the event notification and keepalive support
     - Remove stale travis.yml file
     - Update simptest to truly support PMIx_Abort
     - Some cleanups for client finalize and IOF output
     - Do not forward cached IOF to self
     - Update how C to Python bytes/strings are handled for get and byte objects
     - Add test_v2 directory
     - Add configure logic for RM and sse support
     - Add pstat framework
     - Remove duplicate PMIx_Data_load and PMIx_Data_unload definitions
     - Add manpage files for tools
     - Add sse common component
     - Add the prm components
     - Add the storage framework
     - General update of code base to track master branch
 - PR #2224: common/dstore: Fix inconsistent Makefile.am
 - PR #2216: Cherry-pick updates from master branch
     - Add missing osname endpt elements to bfrops
     - Optimize check for nodes
     - Transfer stdout/err formatting to PMIx
     - Ensure tool output of IOF
 - PR #2208: Cherry-pick updates from master branch
     - Enable re-init of clients
     - Add attribute to indicate copy/nocopy of output directed to files
 - PR #2204: Add a few job error constants
 - PR #2201: Cherry-pick updates from master branch
     - Minor updates based on Standards review
     - Correct references to help-ptl-tool.txt
     - Protect register_nspace against new entries
     - Add oversubscribed attribute
 - PR #2195: configury: Use AC_CHECK_ALIGNOF and fix cross-compiling
 - PR #2190:Cherry-pick updates from master branch
     - Remove duplicative pmdl/ompi directories
     - Add missing m4 file
     - Update hwloc support to handle revised version string
     - Register ompi5 and ompi4 as aliases for ompi plugin
     - Correct the PMIx_Get signature
     - Silence some gcc warnings
     - Silence some gcc11 warnings
     - Protect against bad nspace input
     - Cleanup few lingering gcc11 warnings
     - Protect against duplicate envar harvesting
 - PR #2177: Cleanup shadow variables in dstore base and components
 - PR #2156: Ensure we pass the desired scope on a PMIx_Get call

src/tools/plookup/plookup.c
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2011 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2006-2013 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2009-2012 Cisco Systems, Inc.  All rights reserved.
 * Copyright (c) 2011      Oak Ridge National Labs.  All rights reserved.
 * Copyright (c) 2013-2020 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Mellanox Technologies, Inc.  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 */

#define _GNU_SOURCE
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

#include "include/pmix_tool.h"
#include "src/mca/base/base.h"
#include "src/mca/pinstalldirs/base/base.h"
#include "src/runtime/pmix_rte.h"
#include "src/threads/threads.h"
#include "src/util/cmd_line.h"
#include "src/util/keyval_parse.h"
#include "src/util/show_help.h"

typedef struct {
    pmix_lock_t lock;
    pmix_status_t status;
} mylock_t;

static pmix_proc_t myproc;

/* this is the event notification function we pass down below
 * when registering for general events - i.e.,, the default
 * handler. We don't technically need to register one, but it
 * is usually good practice to catch any events that occur */
static void notification_fn(size_t evhdlr_registration_id, pmix_status_t status,
                            const pmix_proc_t *source, pmix_info_t info[], size_t ninfo,
                            pmix_info_t results[], size_t nresults,
                            pmix_event_notification_cbfunc_fn_t cbfunc, void *cbdata)
{
    /* this example doesn't do anything with default events */
    if (NULL != cbfunc) {
        cbfunc(PMIX_EVENT_ACTION_COMPLETE, NULL, 0, NULL, NULL, cbdata);
    }
}

/* event handler registration is done asynchronously because it
 * may involve the PMIx server registering with the host RM for
 * external events. So we provide a callback function that returns
 * the status of the request (success or an error), plus a numerical index
 * to the registered event. The index is used later on to deregister
 * an event handler - if we don't explicitly deregister it, then the
 * PMIx server will do so when it see us exit */
static void evhandler_reg_callbk(pmix_status_t status, size_t evhandler_ref, void *cbdata)
{
    mylock_t *lock = (mylock_t *) cbdata;

    if (PMIX_SUCCESS != status) {
        fprintf(stderr, "Client %s:%d EVENT HANDLER REGISTRATION FAILED WITH STATUS %d, ref=%lu\n",
                myproc.nspace, myproc.rank, status, (unsigned long) evhandler_ref);
    }
    lock->status = status;
    PMIX_WAKEUP_THREAD(&lock->lock);
}

/*****************************************
 * Global Vars for Command line Arguments
 *****************************************/
typedef struct {
    bool help;
    bool verbose;
    pid_t pid;
    bool wait;
    int timeout;
} pmix_plookup_globals_t;

pmix_plookup_globals_t pmix_plookup_globals = {0};

pmix_cmd_line_init_t cmd_line_opts[]
    = {{NULL, 'h', NULL, "help", 0, &pmix_plookup_globals.help, PMIX_CMD_LINE_TYPE_BOOL,
        "This help message"},

       {NULL, 'v', NULL, "verbose", 0, &pmix_plookup_globals.verbose, PMIX_CMD_LINE_TYPE_BOOL,
        "Be Verbose"},

       {NULL, 'p', NULL, "pid", 1, &pmix_plookup_globals.pid, PMIX_CMD_LINE_TYPE_INT,
        "Specify launcher pid"},

       {NULL, 'w', NULL, "wait", 0, &pmix_plookup_globals.wait, PMIX_CMD_LINE_TYPE_BOOL,
        "Wait for data to be available"},

       {NULL, 't', NULL, "timeout", 0, &pmix_plookup_globals.timeout, PMIX_CMD_LINE_TYPE_INT,
        "Max number of seconds to wait for data to become available"},

       /* End of list */
       {NULL, '\0', NULL, NULL, 0, NULL, PMIX_CMD_LINE_TYPE_NULL, NULL}};

int main(int argc, char **argv)
{
    pmix_status_t rc;
    pmix_info_t *info = NULL;
    size_t ninfo = 0, n;
    mylock_t mylock;
    pmix_cmd_line_t cmd_line;
    pmix_pdata_t *pdata = NULL;
    size_t ndata;
    int count;
    char **keys = NULL;

    /* protect against problems if someone passes us thru a pipe
     * and then abnormally terminates the pipe early */
    signal(SIGPIPE, SIG_IGN);

    /* initialize the output system */
    if (!pmix_output_init()) {
        return PMIX_ERROR;
    }

    /* initialize install dirs code */
    if (PMIX_SUCCESS
        != (rc = pmix_mca_base_framework_open(&pmix_pinstalldirs_base_framework,
                                              PMIX_MCA_BASE_OPEN_DEFAULT))) {
        fprintf(stderr,
                "pmix_pinstalldirs_base_open() failed -- process will likely abort (%s:%d, "
                "returned %d instead of PMIX_SUCCESS)\n",
                __FILE__, __LINE__, rc);
        return rc;
    }
    if (PMIX_SUCCESS != (rc = pmix_pinstall_dirs_base_init(NULL, 0))) {
        fprintf(stderr,
                "pmix_pinstalldirs_base_init() failed -- process will likely abort (%s:%d, "
                "returned %d instead of PMIX_SUCCESS)\n",
                __FILE__, __LINE__, rc);
        return rc;
    }

    /* initialize the help system */
    pmix_show_help_init();

    /* keyval lex-based parser */
    if (PMIX_SUCCESS != (rc = pmix_util_keyval_parse_init())) {
        fprintf(stderr, "pmix_util_keyval_parse_init failed with %d\n", rc);
        return PMIX_ERROR;
    }

    /* Setup the parameter system */
    if (PMIX_SUCCESS != (rc = pmix_mca_base_var_init())) {
        fprintf(stderr, "pmix_mca_base_var_init failed with %d\n", rc);
        return PMIX_ERROR;
    }

    /* register params for pmix */
    if (PMIX_SUCCESS != (rc = pmix_register_params())) {
        fprintf(stderr, "pmix_register_params failed with %d\n", rc);
        return PMIX_ERROR;
    }

    /* Parse the command line options */
    pmix_cmd_line_create(&cmd_line, cmd_line_opts);

    pmix_mca_base_open();
    pmix_mca_base_cmd_line_setup(&cmd_line);
    rc = pmix_cmd_line_parse(&cmd_line, false, false, argc, argv);

    if (PMIX_SUCCESS != rc) {
        if (PMIX_ERR_SILENT != rc) {
            fprintf(stderr, "%s: command line error (%s)\n", argv[0], PMIx_Error_string(rc));
        }
        return rc;
    }

    if (pmix_plookup_globals.help) {
        char *str, *args = NULL;
        args = pmix_cmd_line_get_usage_msg(&cmd_line);
        str = pmix_show_help_string("help-plookup.txt", "usage", true, args);
        if (NULL != str) {
            printf("%s", str);
            free(str);
        }
        free(args);
        /* If we show the help message, that should be all we do */
        exit(0);
    }

    if (pmix_plookup_globals.wait) {
        ++ninfo;
        if (0 < pmix_plookup_globals.timeout) {
            ++ninfo;
        }
    }

    /* determine how many keys were given */
    pmix_cmd_line_get_tail(&cmd_line, &count, &keys);
    if (0 == count) {
        /* must give us at least one key */
        fprintf(stderr, "%s: Must provide at least one key to lookup\n", argv[0]);
        exit(1);
    }
    ndata = count;

    /* if we were given the pid of a starter, then direct that
     * we connect to it */

    /* otherwise, use the system connection first, if available */
    PMIX_INFO_CREATE(info, 1);
    PMIX_INFO_LOAD(&info[0], PMIX_CONNECT_SYSTEM_FIRST, NULL, PMIX_BOOL);
    /* init as a tool */
    if (PMIX_SUCCESS != (rc = PMIx_tool_init(&myproc, info, 1))) {
        fprintf(stderr, "PMIx_tool_init failed: %d\n", rc);
        exit(rc);
    }
    PMIX_INFO_FREE(info, 1);

    /* register a default event handler */
    PMIX_CONSTRUCT_LOCK(&mylock.lock);
    PMIx_Register_event_handler(NULL, 0, NULL, 0, notification_fn, evhandler_reg_callbk,
                                (void *) &mylock);
    PMIX_WAIT_THREAD(&mylock.lock);
    if (PMIX_SUCCESS != mylock.status) {
        fprintf(stderr, "PMIx_Register_event_handler returned bad status: %d\n", rc);
        PMIX_DESTRUCT_LOCK(&mylock.lock);
        goto done;
    }
    PMIX_DESTRUCT_LOCK(&mylock.lock);

    /* setup any info for the lookup */
    if (0 < ninfo) {
        PMIX_INFO_CREATE(info, ninfo);
        PMIX_INFO_LOAD(&info[0], PMIX_WAIT, NULL, PMIX_BOOL);
        if (1 < ninfo) {
            PMIX_INFO_LOAD(&info[1], PMIX_TIMEOUT, &pmix_plookup_globals.timeout, PMIX_INT);
        }
    }

    /* setup the keys */
    PMIX_PDATA_CREATE(pdata, ndata);
    for (n = 0; n < ndata; n++) {
        pmix_strncpy(pdata[n].key, keys[n], PMIX_MAX_KEYLEN);
    }
    /* perform the lookup */
    rc = PMIx_Lookup(pdata, ndata, info, ninfo);

    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "PMIx_Lookup failed: %d\n", rc);
        goto done;
    }

    for (n = 0; n < ndata; n++) {
        fprintf(stderr, "Key: %s\n", pdata[n].key);
    }
    PMIX_PDATA_FREE(pdata, ndata);

done:
    PMIx_tool_finalize();

    return (rc);
}
